astropy/time/core.py

```diff
@@ -2440,7 +2440,12 @@ def _make_array(val, copy=False):
     val : ndarray
         Array version of ``val``.
     """
-    val = np.array(val, copy=copy, subok=True)
+    if isinstance(val, (tuple, list)) and len(val) > 0 and isinstance(val[0], Time):
+        dtype = object
+    else:
+        dtype = None
+
+    val = np.array(val, copy=copy, subok=True, dtype=dtype)
```

Comment:
Can we do this instead with a try/except/catch-warning?  Basically try doing `val = np.array(val, copy=copy, subok=True)` and then if numpy emits a warning or raises an exception then fall through to explicitly specifying `dtype=object`.  That would avoid the special case and hopefully just delegate to numpy whether or not the original `val` is acceptable.

Reply:
@taldcroft - coming back to this PR after a while, I remember why I did the upfront test: I'm quite happy to just do try/except for the exception, but changing the warnings filter seems slow and excessive. So, the question then is whether we mind the warnings for the case of `[scalar_time, array_time]`.

p.s. Ideally, to me, the case of an iterable of time would never get turned into an array in the first place...

Reply:
I did a little profiling, and the impact on initialization of catching warnings is not huge, probably an increase of 2 us on a typical initialization of 65 us.  The catching bit could be removed once versions of numpy that issue the deprecation are no longer supported by astropy.

Overall I agree that the current API of allowing a list of ragged `Time` objects and concatenating is odd and in fact I was surprised myself at that behavior.  :smile:.  But that's a separate decision about whether to deprecate that as well.

I think that allowing a (non-ragged) iterable of Time is OK, certainly given the current behavior:
```
>>> t = Time([1,2,3], format='cxcsec')
>>> np.array(t)
array([<Time object: scale='tt' format='cxcsec' value=1.0>,
       <Time object: scale='tt' format='cxcsec' value=2.0>,
       <Time object: scale='tt' format='cxcsec' value=3.0>], dtype=object)
>>> Time(np.array(t))
<Time object: scale='tt' format='cxcsec' value=[1. 2. 3.]> 
# Round trips!  And would fail your current special-case test
```
## Timing

```python
import warnings
from astropy.time import Time
```


```python
def fxn(**kwargs):
    x = np.array([1], **kwargs)

def do_fxn():
    with warnings.catch_warnings(record=True) as w:
        try:
            fxn()
        except TypeError:
            fxn(dtype=object)
            
        if len(w) > 0:
            # Check for the automatic object array deprecation warning and ignore,
            # otherwise emit the warning.
            pass
```


```python
timeit do_fxn()
```

    2.73 µs ± 31.9 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)



```python
timeit fxn()
```

    667 ns ± 14.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)



```python
timeit Time(50000, format='mjd')
```

    63 µs ± 1.9 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)



```python
timeit Time('2019-01-01', format='iso')
```

    66.5 µs ± 1.37 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

Reply:
That might work, though it means the type of any other warning is lost (only keeps the warning message), and we add yet more that is not thread-safe. The logical route here would be not to record but set up an appropriate filter instead, but that adds another few us, while my `if` statement is just one line and well under a micro-second.

Reply:
OK, fair enough.  I do think that this kind of special-case opt-in catching is fragile and has some risk of failing for an oddball case you haven't thought about.  But I can't decide if that is worse or doing just try/except and exposing annoying warnings that will be obscure to the user.  So I'll let you decide.  :smile: